Add Plant trait that lets you "commune" with trees

thquinn opened the pull request:
#### Summary
```SUMMARY: Content "Add Plant trait that lets you commune with trees"```

#### Purpose of change
I read [this article](https://e360.yale.edu/features/exploring_how_and_why_trees_talk_to_each_other) positing that trees communicate with each other and figured... might be cool?

#### Describe the solution
I added "Mycorrhizal Communion," a post-threshold trait to Plant. Activating it causes you to enter a meditative state that slowly reveals the overmap along contiguous forest tiles.

![ss](https://user-images.githubusercontent.com/1022438/55781695-c76e9800-5a78-11e9-9826-88f4b81e79ac.png)

Every minute, it performs a breadth-first search to find the nearest connected forest that has unrevealed overmap tiles near it and reveals those tiles. The communion ends once the player has seen everything there is to see. Hunger, thirst, and fatigue rates are halved during the communion.

#### Could use some advice about
* Is there a better way to check if an overmap tile is forest? I assume it's not string comparing to "forest", which is what I'm doing. Ideally it would include forest trails and maybe swamps since they got trees in 'em too.
* Any activity end conditions or player messages I should add?
* Any other effects/behaviors anyone can think of for communing with trees?

Maddremor commented:
On one hand: This seems too much like magic to fit as a mutation IMO. How would it work conceptually?
On the other hand: This is a **really** cool mechanic. You could also make something similar with fish and rivers.

thquinn commented:
@Maddremor The linked article explains a bit of how the real-life mechanism supposedly operates. [Here's a National Geographic video](https://www.youtube.com/watch?v=7kHZ0a_6TxY) that also covers it. It seems pretty speculative to me, but I'm no expert.

Rail-Runner commented:
A few thoughts about this, assuming it does somehow work: trees don't really have "eyes" of any kind and likely wouldn't be able to sense a building that is close to a forest (or just know about it as much as "no trees grow there" and surely not the fact that the building is actually a lab or something else; trees can't understand technology-related things at all), so it should only reveal forest tiles exactly.

Also, this communication system would very likely need roots to be usable, so it's probably only possible for plant mutants, and should require player to mutate some kind of roots first.

Zireael07 commented:
+1 to restricting it to Plant and requiring Roots first.

thquinn commented:
@Rail-Runner The trees might not understand technology or man-made structures, but as a human, you'd be able to take the information you get and extract more meaning from it than what the trees "know." Could you tell what kind of obstacles were out there? Maybe not, but from a gameplay perspective it seems fair.

Making it Plant-only and requiring roots makes sense.

AMurkin commented:
> Is there a better way to check if an overmap tile is forest? I assume it's not string comparing to "forest", which is what I'm doing. Ideally it would include forest trails and maybe swamps since they got trees in 'em too.

In #28761 this is done the same way.

Fris0uman commented:
Mycorrhiza are fungi, "Mycorrhizal Communion" implies that you need a a symbiotic fungus that is linking your roots to the ones of the next tree. Developing such a symbiosis would require for the player to stay rooted for a long time, maybe a few days or weeks. So it would maybe be better to rename it to "Plant communication" and implie that it uses an [Haustorium](https://en.wikipedia.org/wiki/Haustorium) for example. Haustorium don't nenecessarily grow from roots.

Also I don't think it should be able to reveal overmap that are not directly adjacent to a forest.

thquinn commented:
> Developing such a symbiosis would require for the player to stay rooted for a long time, maybe a few days or weeks.

I could see adding a startup time, but what are you basing this claim on? The chemical messages — according to the sources I've posted, anyway — are already being sent, it could be as easy as settling your roots in and intercepting them.

> So it would maybe be better to rename it to "Plant communication" and implie that it uses an [Haustorium](https://en.wikipedia.org/wiki/Haustorium) for example.

I'm basing this implementation on the sources I can find on mycorrhizae, and I'm not seeing similar claims about haustoria.

Fris0uman commented:
So from what I remember and what I understand of the article the communication process requires the molecules to be carried throught the fungi network (even more so if we want to uncover over map tiles far away). So if you want to get the message  you need to be plugged into the network and not just have your roots around.

> what are you basing this claim on?

I don't have readily avaible articles right now but the idea is : 
- This communication is a chemical one so it's limited by the speed of diffusion, the molecules do have to travel the distance to carry the message and it's not as fast a electrical stuff.
- It requires a symbiosis with the fungi network which require cell growth and division on both side and this takes time.

But in the end we just need to decide where to start the suspension of disbelief, maybe we don't need to bother with anything I said and just tweak the time it takes and the amount of map revealed.

SantosValourion commented:
Maybe theres need to be a trait to give elf-a roots in there hair, then it could lead to this trait. Just a cool idea im picturing in my head, the elf uses there hair roots to talk to the trees...

thquinn commented:
* Added a startup time: 60-90 minutes for `ROOTS2`, 15-30 minutes for `ROOTS3`.
* Reduced vision radius from 3 to 2.
* Now searches through forests AND swamps.
* Can only be started from a forest or swamp.
* Added snippets.

> But in the end we just need to decide where to start the suspension of disbelief, maybe we don't need to bother with anything I said and just tweak the time it takes and the amount of map revealed.

Yeah, I'm curious to hear where people want to draw that line.

codemime commented:
> Is there a better way to check if an overmap tile is forest? I assume it's not string comparing to "forest", which is what I'm doing. Ideally it would include forest trails and maybe swamps since they got trees in 'em too.

You could use `overmap_location` for this. That's slightly better than specific omap terrain types, though still requires string comparison to work.

kevingranade commented:
I'm not too concerned about the justification, it's off in mutation magic land to start with, and who's to say (ok me I guess) that the trees aren't mutated too.

I do think that having it reveal buildings is a bit much. How about it does radius 3, but does not reveal anything but trees and fields.  the player could still infer the presence of roads and buildings from oddities in the map shape without it implying that trees can detect the difference between a house and a grocery store.

Little-119 commented:
I think only revealing forest tiles themselves would be the most useful, while keeping the trait useful as it still allows the player to infer the presence of roads and buildings from gaps in the forests, as kevingranade said.

thquinn commented:
* Added `WOODED` overmap_terrain flag and put it on several terrain types.
* Added filtered overmapbuffer::reveal function, using that to only reveal wooded terrain and fields.
* Radius back to 3.

Before communion:
![ss](https://user-images.githubusercontent.com/1022438/55850703-7ff11600-5b23-11e9-866d-e364151b49f6.png)

After 12 hours:
![ss2](https://user-images.githubusercontent.com/1022438/55850711-84b5ca00-5b23-11e9-838d-5d20222ef4c7.png)

Fris0uman commented:
I must admit this is quite cool. 👍 

Two randoms ideas :
 - It could have a second level that let's you uncover what's 1z level below the `WOODED` tile
 - A chance to get fungal infection if there's fungus/Fungal tower/Fungal Bloom on the tile discovered ? (might be a bit too much)

thquinn commented:
Now adds +1 morale/minute while communing, up to 15.

> It could have a second level that let's you uncover what's 1z level below the `WOODED` tile

Most of these wooded tiles don't have anything under them, I don't think?

> A chance to get fungal infection if there's fungus/Fungal tower/Fungal Bloom on the tile discovered ? (might be a bit too much)

There's definitely room for interactions with fungal stuff, it being right there in the name and all. Since the ability isn't super useful right now anyway, though, I think it would be cool to add more incidentally positive effects. I added a small morale boost for starters.

I-am-Erk commented:
I think this is a really nice, flavourful ability that is extremely useful without being too powerful. We need more of this kind of thing.

I'd avoid having fungal infection be a possibility, it would just be a pain in the ass since you don't get much choice and you have no way to determine in advance what the risk of finding a fungal infection is. Instead if there's interaction with fungus I'd suggest it be that fungal areas form a blank spot in the map since any root signalling would be defunct there.

thquinn commented:
@I-am-Erk Thanks! I do think it feels a little flat right now, as it only provides three things:
* overmap vision
* morale
* partial life support

and doesn't really live up to the "spiritual" side of the idea. I'm not totally sure how to represent that in gameplay terms, though; it would probably just be an accumulation of minor additions over time.

Ilysen commented:
Hey, speaking of spiritual, making the morale bonus bigger and longer if you have the Spiritual trait might be cool! I'm not entirely sure it's very easy to represent it otherwise, though.

I-am-Erk commented:
Bear in mind this is a single mutation. It doesn't need to be an entire gameplay mechanic on its own... In fact that's probably too much for a single mutation

thquinn commented:
@Ilysen Great idea, done.

@I-am-Erk True.

thquinn commented:
@ifreund Switched over to `overmap_land_use_code_id`. It works, but these categories aren't as broad as I'd hoped they would be... in overmap_terrain_abstract.json, we've got `junkyard`, `orchard` AND `nursery`, `cemetery`. I'm already having to check against 4 IDs for the results I want, and I worry that new types of forests that get implemented in the future are going to make up their own new IDs instead of just using "forest." Not that my `WOODED` tag was any more future-proof, but... maybe I'm not fully appreciating the usefulness of land use codes.

kevingranade commented:
Works great, but balance wise it's discovering a lot of open field area that doesn't seem intended, and it's finding disconnected forest tiles.

Rail-Runner commented:
I think it's because those areas were known already before mutation was used.

thquinn commented:
That's right. Here's an example of 12 hours of communion with nothing revealed prior:
![ss](https://user-images.githubusercontent.com/1022438/56189916-1c6d5980-5ff7-11e9-9200-272fcddcc2ff.png)

and here's what the revealed overmap looks like:
![ss2](https://user-images.githubusercontent.com/1022438/56189939-298a4880-5ff7-11e9-9387-90f3cc050a31.png)

I-am-Erk commented:
> @ifreund Switched over to `overmap_land_use_code_id`. It works, but these categories aren't as broad as I'd hoped they would be... in overmap_terrain_abstract.json, we've got `junkyard`, `orchard` AND `nursery`, `cemetery`. I'm already having to check against 4 IDs for the results I want, and I worry that new types of forests that get implemented in the future are going to make up their own new IDs instead of just using "forest." Not that my `WOODED` tag was any more future-proof, but... maybe I'm not fully appreciating the usefulness of land use codes.

Land use codes should be pretty stable here on out: they aren't something we thought up, but official land use codes from Massachusetts GIS data.

ifreund commented on src/activity_handlers.cpp:
```diff
@@ -3337,3 +3341,63 @@ void activity_handlers::robot_control_finish( player_activity *act, player *p )
+    q.push( loc );
+    seen.insert( loc );
+    const std::function<bool( const oter_id & )> filter = []( const oter_id & ter ) {
+        const overmap_land_use_code_id forest_id = overmap_land_use_code_id( "forest" );
```
Unnecessary line?
```suggestion
```

ifreund commented on src/mutation.cpp:
```diff
@@ -501,6 +505,35 @@ void player::activate_mutation( const trait_id &mut )
         return;
+    } else if( mut == trait_TREE_COMMUNION ) {
+        tdata.powered = false;
+        const overmap_land_use_code_id forest_id = overmap_land_use_code_id( "forest" );
```
This one too.
```suggestion
```

thquinn commented:
@I-am-Erk Ah, good to know. It's reassuring that there's a fixed source they're drawn from.

@ifreund No problem at all. Thanks for the reviews!
